File: pfaa-cli/pfaa_bridge.py
#!/usr/bin/env python3
"""
PFAA Bridge — Python-side JSON-over-stdin/stdout bridge.

This is the entry point that the Node.js CLI spawns as a subprocess.
It receives JSON commands on stdin and returns JSON responses on stdout.

The bridge exposes the full PFAA engine:
- Tool execution (27+ tools, phase-aware)
- Goal decomposition and autonomous execution
- Memory operations (5-layer meta-learning)
- Claude Code integration
- Self-build cycles
- Benchmarks

Protocol:
  → {"action": "execute_tool", "args": {"name": "compute", "args": ["sqrt(42)"]}, "id": "cmd_1"}
  ← {"id": "cmd_1", "success": true, "data": {...}, "elapsed_us": 42, "phase": "VAPOR"}

Python 3.15: lazy import, frozendict, kqueue subprocess.
"""

from __future__ import annotations

import asyncio
import sys
import os
import time

lazy import json
lazy import traceback

# Add parent directory to path for PFAA engine imports
sys.path.insert(0, os.path.dirname(os.path.dirname(os.path.abspath(__file__))))


async def handle_command(cmd: dict) -> dict:
    """Route a bridge command to the appropriate PFAA engine function."""
    action = cmd.get("action", "")
    args = cmd.get("args", {})
    cmd_id = cmd.get("id", "unknown")
    start_ns = time.perf_counter_ns()

    try:
        if action == "status":
            from agent_setup_cli.core.framework import Framework
            fw = Framework()
            data = fw.status()
            return _ok(cmd_id, data, start_ns)

        elif action == "list_tools":
            from agent_setup_cli.core.tools import ToolRegistry
            import agent_setup_cli.core.tools_extended  # noqa: F401
            try:
                import agent_setup_cli.core.tools_generated  # noqa: F401
            except ImportError:
                pass

            registry = ToolRegistry.get()
            tools = [
                {
                    "name": t.name,
                    "phase": t.phase.name,
                    "description": t.description,
                    "capabilities": list(t.capabilities),
                    "isolated": t.isolated,
                }
                for t in registry.list_tools()
            ]
            return _ok(cmd_id, tools, start_ns)

        elif action == "execute_tool":
            from agent_setup_cli.core.tools import ToolRegistry
            from agent_setup_cli.core.persistence import PersistentMemory
            import agent_setup_cli.core.tools_extended  # noqa: F401

            registry = ToolRegistry.get()
            mem = PersistentMemory()

            tool_name = args.get("name", "")
            tool_args = tuple(args.get("args", []))

            result = await registry.execute(tool_name, *tool_args)
            mem.record(result, tool_name, tool_args)
            mem.close()

            return _ok(
                cmd_id,
                result.result,
                start_ns,
                phase=result.phase_used.name,
                elapsed_us=result.elapsed_us,
            )

        elif action == "run_goal":
            from agent_setup_cli.core.framework import Framework
            fw = Framework()
            goal = args.get("goal", "")
            state = await fw.run(goal)
            data = {
                "goal_id": state.goal_id,
                "status": state.status.value,
                "subtasks": [
                    {
                        "description": st.description,
                        "status": st.status,
                        "tool": st.tool,
                    }
                    for st in state.subtasks
                ],
                "completed": sum(1 for st in state.subtasks if st.status == "completed"),
                "failed": sum(1 for st in state.subtasks if st.status == "failed"),
            }
            await fw.shutdown()
            return _ok(cmd_id, data, start_ns)

        elif action == "scatter":
            from agent_setup_cli.core.tools import ToolRegistry
            import agent_setup_cli.core.tools_extended  # noqa: F401

            registry = ToolRegistry.get()
            tool_name = args.get("tool", "")
            inputs = args.get("inputs", [])

            results = await registry.execute_many([
                (tool_name, (inp,), {}) for inp in inputs
            ])

            data = []
            for inp, result in zip(inputs, results):
                if isinstance(result, Exception):
                    data.append({"input": inp, "success": False, "error": str(result)})
                else:
                    data.append({
                        "input": inp,
                        "success": True,
                        "result": result.result,
                        "phase": result.phase_used.name,
                        "elapsed_us": result.elapsed_us,
                    })

            return _ok(cmd_id, data, start_ns)

        elif action == "memory_status":
            from agent_setup_cli.core.persistence import PersistentMemory
            mem = PersistentMemory()
            status = mem.status()
            mem.close()
            return _ok(cmd_id, {
                "l1Episodes": status["l1_episodes"],
                "l2Patterns": status["l2_patterns"],
                "l3Strategies": status["l3_strategies"],
                "l4LearningRate": status["l4_learning_rate"],
                "l5Knowledge": status["l5_knowledge"],
                "dbSizeKb": status["db_size_kb"],
            }, start_ns)

        elif action == "force_learn":
            from agent_setup_cli.core.persistence import PersistentMemory
            mem = PersistentMemory()
            mem.force_learn()
            mem.close()
            return _ok(cmd_id, {"learned": True}, start_ns)

        elif action == "ask_claude":
            from agent_setup_cli.core.claude_bridge import ClaudeBridge, ClaudeConfig
            bridge = ClaudeBridge(config=ClaudeConfig(
                model=args.get("model", "sonnet"),
            ))
            result = await bridge.ask(args.get("prompt", ""))
            return _ok(cmd_id, {
                "success": result.success,
                "output": result.output,
                "elapsedMs": result.elapsed_ms,
            }, start_ns)

        elif action == "generate_code":
            from agent_setup_cli.core.claude_bridge import ClaudeBridge, ClaudeConfig
            bridge = ClaudeBridge(config=ClaudeConfig())
            result = await bridge.generate_code(
                description=args.get("description", ""),
                language=args.get("language", "python"),
                output_file=args.get("output_file"),
            )
            return _ok(cmd_id, {
                "success": result.success,
                "code": result.output,
                "file": args.get("output_file"),
            }, start_ns)

        elif action == "list_checkpoints":
            from agent_setup_cli.core.autonomous import AutonomousAgent
            agent = AutonomousAgent()
            checkpoints = agent.checkpoints()
            return _ok(cmd_id, checkpoints, start_ns)

        elif action == "resume_goal":
            from agent_setup_cli.core.autonomous import AutonomousAgent
            agent = AutonomousAgent()
            goal_id = args.get("goal_id", "")
            state = await agent.resume(goal_id)
            return _ok(cmd_id, {
                "goal_id": state.goal_id,
                "status": state.status.value,
                "completed": sum(1 for st in state.subtasks if st.status == "completed"),
            }, start_ns)

        elif action == "get_memory":
            from agent_setup_cli.core.persistence import PersistentMemory
            mem = PersistentMemory()
            dump = mem.dump()
            mem.close()
            return _ok(cmd_id, dump, start_ns)

        elif action == "pipeline":
            from agent_setup_cli.core.tools import ToolRegistry
            import agent_setup_cli.core.tools_extended  # noqa: F401

            registry = ToolRegistry.get()
            steps = args.get("steps", [])
            results = []
            for step in steps:
                tool_name = step.get("tool", "")
                tool_args = tuple(step.get("args", []))
                try:
                    result = await registry.execute(tool_name, *tool_args)
                    results.append({
                        "tool": tool_name,
                        "success": True,
                        "result": result.result,
                        "phase": result.phase_used.name,
                        "elapsed_us": result.elapsed_us,
                    })
                except Exception as e:
                    results.append({
                        "tool": tool_name,
                        "success": False,
                        "result": str(e),
                        "phase": "ERROR",
                        "elapsed_us": 0,
                    })
            return _ok(cmd_id, results, start_ns)

        elif action == "explore":
            from agent_setup_cli.core.framework import Framework
            fw = Framework()
            rounds = args.get("rounds", 200)
            epsilon = args.get("epsilon", 0.3)
            # Run exploration
            results = {}
            for i in range(rounds):
                for tool_name in ["compute", "hash_data", "json_parse", "regex_extract", "line_count"]:
                    try:
                        result = await fw.tool(tool_name, "test")
                        results.setdefault(tool_name, []).append({
                            "phase": result.phase_used.name,
                            "elapsed_us": result.elapsed_us,
                        })
                    except Exception:
                        pass
            # Summarize
            summary = {}
            for tool_name, runs in results.items():
                by_phase = {}
                for r in runs:
                    by_phase.setdefault(r["phase"], []).append(r["elapsed_us"])
                summary[tool_name] = {
                    phase: {"avg_us": sum(times) // len(times), "count": len(times)}
                    for phase, times in by_phase.items()
                }
            await fw.shutdown()
            return _ok(cmd_id, {"rounds": rounds, "epsilon": epsilon, "tools": summary}, start_ns)

        elif action == "spawn_team":
            import subprocess as sp
            goal = args.get("goal", "")
            mode = args.get("mode", "basic")
            script = "agents/team/remix_spawn.py" if mode == "remix" else "agents/team/spawn.py"
            script_path = os.path.join(os.path.dirname(os.path.dirname(os.path.abspath(__file__))), script)

            if not os.path.exists(script_path):
                return _err(cmd_id, f"Team script not found: {script_path}", start_ns)

            proc = sp.run(
                [sys.executable, script_path, goal],
                capture_output=True, text=True, timeout=600,
                cwd=os.path.dirname(os.path.dirname(os.path.abspath(__file__))),
            )
            return _ok(cmd_id, {
                "mode": mode,
                "goal": goal,
                "exit_code": proc.returncode,
                "stdout": proc.stdout[-2000:] if proc.stdout else "",
                "stderr": proc.stderr[-1000:] if proc.stderr else "",
            }, start_ns)

        elif action == "self_build":
            from agent_setup_cli.core.self_build import SelfBuilder
            builder = SelfBuilder()
            try:
                result = await builder.build_self(
                    auto_apply=args.get("auto_apply", False),
                )
                return _ok(cmd_id, {
                    k: v for k, v in result.items() if k != "proposals"
                }, start_ns)
            finally:
                await builder.shutdown()

        elif action == "benchmark":
            from agent_setup_cli.core.benchmark import main as bench_main
            # Capture benchmark output
            import io
            old_stdout = sys.stdout
            sys.stdout = io.StringIO()
            try:
                await bench_main()
                output = sys.stdout.getvalue()
            finally:
                sys.stdout = old_stdout
            return _ok(cmd_id, {"output": output}, start_ns)

        else:
            return _err(cmd_id, f"Unknown action: {action}", start_ns)

    except Exception as e:
        return _err(cmd_id, f"{type(e).__name__}: {e}", start_ns)


def _ok(cmd_id: str, data: any, start_ns: int, **extra) -> dict:
    elapsed_us = (time.perf_counter_ns() - start_ns) // 1000
    return {"id": cmd_id, "success": True, "data": data, "elapsed_us": elapsed_us, **extra}


def _err(cmd_id: str, error: str, start_ns: int) -> dict:
    elapsed_us = (time.perf_counter_ns() - start_ns) // 1000
    return {"id": cmd_id, "success": False, "data": None, "error": error, "elapsed_us": elapsed_us}


async def main():
    """Main bridge loop — read JSON commands from stdin, write responses to stdout."""
    # Signal ready
    sys.stdout.write(json.dumps({"id": "ready", "success": True, "data": "PFAA bridge ready"}) + "\n")
    sys.stdout.flush()

    loop = asyncio.get_event_loop()
    reader = asyncio.StreamReader()
    protocol = asyncio.StreamReaderProtocol(reader)
    await loop.connect_read_pipe(lambda: protocol, sys.stdin)

    while True:
        try:
            line = await reader.readline()
            if not line:
                break  # EOF

            line_str = line.decode("utf-8").strip()
            if not line_str:
                continue

            cmd = json.loads(line_str)
            response = await handle_command(cmd)

            sys.stdout.write(json.dumps(response, default=str) + "\n")
            sys.stdout.flush()

        except json.JSONDecodeError as e:
            sys.stderr.write(f"Invalid JSON: {e}\n")
            sys.stderr.flush()
        except Exception as e:
            sys.stderr.write(f"Bridge error: {e}\n")
            sys.stderr.flush()
            sys.stdout.write(json.dumps({
                "id": "error",
                "success": False,
                "error": str(e),
            }) + "\n")
            sys.stdout.flush()


if __name__ == "__main__":
    asyncio.run(main())

File: pfaa-cli/src/bridge/pfaa-bridge.ts
/**
 * Aussie Agents Bridge — Node.js ↔ Python 3.15 Engine Bridge
 *
 * Spawns the Aussie Agents Python engine as a subprocess and communicates
 * via JSON-over-stdin/stdout. Supports streaming results, phase
 * transitions, and memory synchronization.
 *
 * This is the core integration point: the Node.js CLI orchestrates
 * while the Python 3.15 engine does the heavy lifting.
 */

import { spawn, type ChildProcess } from 'node:child_process';
import { join } from 'node:path';
import { existsSync } from 'node:fs';
import { EventEmitter } from 'node:events';
import { getLogger } from '../utils/logger.js';
import type {
  Phase,
  ToolResult,
  MemoryStatus,
  AgentResult,
  AgentRole,
  StreamEvent,
  EventType,
} from '../types.js';

const log = getLogger('bridge');

export interface BridgeConfig {
  pythonPath: string;
  enginePath: string;
  workingDir: string;
  timeoutMs: number;
  maxConcurrent: number;
  /** Max milliseconds to wait for the Python engine to become ready (default 30s). */
  startupTimeoutMs: number;
}

interface BridgeCommand {
  action: string;
  args: Record<string, unknown>;
  id: string;
}

interface BridgeResponse {
  id: string;
  success: boolean;
  data: unknown;
  error?: string;
  elapsed_us?: number;
  phase?: Phase;
}

export class PFAABridge extends EventEmitter {
  private config: BridgeConfig;
  private process: ChildProcess | null = null;
  private pending = new Map<string, {
    resolve: (val: BridgeResponse) => void;
    reject: (err: Error) => void;
    timer: ReturnType<typeof setTimeout>;
  }>();
  private buffer = '';
  private ready = false;
  private commandId = 0;

  constructor(config: BridgeConfig) {
    super();
    this.config = config;
  }

  async start(): Promise<void> {
    if (this.process) return;

    const engineEntry = this.findEngineEntry();
    if (!engineEntry) {
      throw new Error(
        `Aussie Agents engine not found. Expected at ${this.config.enginePath}. ` +
        'Run: pip install -e . in the pfaa-engine root.',
      );
    }

    log.info('Starting Aussie Agents engine', {
      python: this.config.pythonPath,
      engine: engineEntry,
    });

    this.process = spawn(this.config.pythonPath, ['-u', engineEntry], {
      cwd: this.config.workingDir,
      stdio: ['pipe', 'pipe', 'pipe'],
      env: {
        ...process.env,
        PYTHONUNBUFFERED: '1',
        PFAA_MODE: 'bridge',
      },
    });

    this.process.stdout?.on('data', (chunk: Buffer) => {
      this.buffer += chunk.toString();
      this.processBuffer();
    });

    this.process.stderr?.on('data', (chunk: Buffer) => {
      const msg = chunk.toString().trim();
      if (msg) log.debug(`[engine] ${msg}`);
    });

    this.process.on('exit', (code) => {
      log.warn('Aussie Agents engine exited', { code });
      this.ready = false;
      this.process = null;
      this.rejectAll(new Error(`Engine exited with code ${code}`));
      this.emit('exit', code);
    });

    this.process.on('error', (err) => {
      log.error('Engine process error', { error: err.message });
      this.emit('error', err);
    });

    // Wait for the engine to signal readiness (or timeout)
    await new Promise<void>((resolve, reject) => {
      const startupTimeout = setTimeout(() => {
        reject(new Error(
          `Engine failed to become ready within ${this.config.startupTimeoutMs}ms. ` +
          'Check that the Python engine is installed and pfaa_bridge.py prints a ready message.',
        ));
      }, this.config.startupTimeoutMs);

      const onData = (chunk: Buffer) => {
        const text = chunk.toString();
        // Engine sends a JSON line with "ready" or we see any valid JSON response
        if (text.includes('"ready"') || text.includes('"status"')) {
          clearTimeout(startupTimeout);
          this.process?.stdout?.off('data', onData);
          resolve();
        }
      };

      // If the process exits before ready, reject immediately
      const onExit = (code: number | null) => {
        clearTimeout(startupTimeout);
        this.process?.stdout?.off('data', onData);
        reject(new Error(`Engine exited with code ${code} before becoming ready`));
      };

      this.process!.stdout?.on('data', onData);
      this.process!.once('exit', onExit);

      // Also resolve after the timeout if stdout is already flowing
      // (covers engines that don't send an explicit ready message)
    });

    this.ready = true;
    log.info('Aussie Agents engine bridge started');
  }

  async stop(): Promise<void> {
    if (!this.process) return;
    this.process.kill('SIGTERM');
    this.process = null;
    this.ready = false;
    this.rejectAll(new Error('Bridge stopped'));
  }

  get isRunning(): boolean {
    return this.ready && this.process !== null;
  }

  // ── Engine Commands ──────────────────────────────────────────────

  async status(): Promise<MemoryStatus & { tools: number; uptime_ms: number }> {
    const resp = await this.send('status', {});
    return resp.data as MemoryStatus & { tools: number; uptime_ms: number };
  }

  async listTools(): Promise<Array<{
    name: string;
    phase: Phase;
    description: string;
    capabilities: string[];
  }>> {
    const resp = await this.send('list_tools', {});
    return (resp.data ?? []) as Array<{ name: string; phase: Phase; description: string; capabilities: string[] }>;
  }

  async executeTool(name: string, ...args: unknown[]): Promise<ToolResult> {
    const resp = await this.send('execute_tool', { name, args });
    return {
      tool: name,
      success: resp.success,
      result: resp.data,
      phaseUsed: resp.phase || ('VAPOR' as Phase),
      elapsedUs: resp.elapsed_us || 0,
    };
  }

  async runGoal(goal: string): Promise<AgentResult> {
    const resp = await this.send('run_goal', { goal }, this.config.timeoutMs);
    return resp.data as AgentResult;
  }

  async scatter(
    toolName: string,
    inputs: unknown[],
  ): Promise<ToolResult[]> {
    const resp = await this.send('scatter', { tool: toolName, inputs });
    return resp.data as ToolResult[];
  }

  async selfBuild(autoApply: boolean = false): Promise<Record<string, unknown>> {
    const resp = await this.send('self_build', { auto_apply: autoApply }, 300_000);
    return resp.data as Record<string, unknown>;
  }

  async benchmark(): Promise<Record<string, unknown>> {
    const resp = await this.send('benchmark', {}, 600_000);
    return resp.data as Record<string, unknown>;
  }

  // ── Goal & Task Management ──────────────────────────────────────

  async listCheckpoints(): Promise<Array<{
    goal_id: string;
    goal: string;
    status: string;
    subtasks: number;
    completed: number;
  }>> {
    const resp = await this.send('list_checkpoints', {});
    return (resp.data ?? []) as Array<{ goal_id: string; goal: string; status: string; subtasks: number; completed: number }>;
  }

  async resumeGoal(goalId: string): Promise<AgentResult> {
    const resp = await this.send('resume_goal', { goal_id: goalId }, this.config.timeoutMs);
    return resp.data as AgentResult;
  }

  // ── Memory Operations ─────────────────────────────────────────

  async getMemory(): Promise<{
    patterns: Array<{ tool: string; best_phase: string; avg_latency_us: number; confidence: number }>;
    strategies: Array<{ tool: string; from_phase: string; to_phase: string; speedup: string }>;
    episodes: number;
    knowledge: Array<{ pattern: string; frequency: number }>;
  }> {
    const resp = await this.send('get_memory', {});
    return resp.data as { patterns: Array<{ tool: string; best_phase: string; avg_latency_us: number; confidence: number }>; strategies: Array<{ tool: string; from_phase: string; to_phase: string; speedup: string }>; episodes: number; knowledge: Array<{ pattern: string; frequency: number }> };
  }

  async forceLearn(): Promise<{ learned: boolean }> {
    const resp = await this.send('force_learn', {});
    return resp.data as { learned: boolean };
  }

  // ── Scatter / Pipeline ────────────────────────────────────────

  async pipeline(steps: Array<{ tool: string; args: unknown[] }>): Promise<Array<{
    tool: string;
    success: boolean;
    result: unknown;
    phase: string;
    elapsed_us: number;
  }>> {
    const resp = await this.send('pipeline', { steps }, this.config.timeoutMs);
    return (resp.data ?? []) as Array<{ tool: string; success: boolean; result: unknown; phase: string; elapsed_us: number }>;
  }

  // ── Exploration ───────────────────────────────────────────────

  async explore(rounds: number = 200, epsilon: number = 0.3): Promise<Record<string, unknown>> {
    const resp = await this.send('explore', { rounds, epsilon }, 300_000);
    return resp.data as Record<string, unknown>;
  }

  // ── Team Spawning ─────────────────────────────────────────────

  async spawnTeam(goal: string, mode: 'basic' | 'remix' = 'basic'): Promise<Record<string, unknown>> {
    const resp = await this.send('spawn_team', { goal, mode }, 600_000);
    return resp.data as Record<string, unknown>;
  }

  // ── Claude Integration ───────────────────────────────────────────

  async askClaude(prompt: string, model?: string): Promise<{
    success: boolean;
    output: string;
    elapsedMs: number;
  }> {
    const resp = await this.send('ask_claude', { prompt, model }, 120_000);
    return resp.data as { success: boolean; output: string; elapsedMs: number };
  }

  async generateCode(
    description: string,
    language: string = 'python',
    outputFile?: string,
  ): Promise<{ success: boolean; code: string; file?: string }> {
    const resp = await this.send('generate_code', {
      description,
      language,
      output_file: outputFile,
    }, 120_000);
    return resp.data as { success: boolean; code: string; file?: string };
  }

  // ── Session Persistence ──────────────────────────────────────────

  async saveSession(sessionId: string, state: Record<string, unknown>): Promise<{saved: string}> {
    const resp = await this.send('save_session', { session_id: sessionId, state });
    return resp.data as {saved: string};
  }

  async loadSession(sessionId?: string): Promise<Record<string, unknown>> {
    const resp = await this.send('load_session', sessionId ? { session_id: sessionId } : {});
    return resp.data as Record<string, unknown>;
  }

  // ── Deferred Tool Discovery ──────────────────────────────────────

  async searchTools(query: string, limit: number = 5): Promise<Array<{
    name: string;
    description: string;
    phase: string;
    capabilities: string[];
  }>> {
    const resp = await this.send('deferred_tool_search', { query, limit });
    return (resp.data ?? []) as Array<{ name: string; description: string; phase: string; capabilities: string[] }>;
  }

  // ── Internal ─────────────────────────────────────────────────────

  private async send(
    action: string,
    args: Record<string, unknown>,
    timeoutMs?: number,
  ): Promise<BridgeResponse> {
    if (!this.ready || !this.process?.stdin) {
      throw new Error('Bridge not started. Call bridge.start() first.');
    }

    const id = `cmd_${++this.commandId}`;
    const cmd: BridgeCommand = { action, args, id };

    return new Promise((resolve, reject) => {
      const timer = setTimeout(() => {
        this.pending.delete(id);
        reject(new Error(`Bridge command '${action}' timed out after ${timeoutMs || this.config.timeoutMs}ms`));
      }, timeoutMs || this.config.timeoutMs);

      this.pending.set(id, { resolve, reject, timer });
      try {
        this.process!.stdin!.write(JSON.stringify(cmd) + '\n');
      } catch (err) {
        this.pending.delete(id);
        clearTimeout(timer);
        reject(new Error(`Bridge stdin write failed: ${err}`));
      }
    });
  }

  private processBuffer(): void {
    const lines = this.buffer.split('\n');
    this.buffer = lines.pop() || '';

    for (const line of lines) {
      if (!line.trim()) continue;
      try {
        const resp = JSON.parse(line) as BridgeResponse;
        if (resp.id && this.pending.has(resp.id)) {
          const { resolve, timer } = this.pending.get(resp.id)!;
          clearTimeout(timer);
          this.pending.delete(resp.id);
          resolve(resp);
        } else {
          // Streaming event from engine
          this.emit('event', resp);
        }
      } catch {
        log.trace('Non-JSON from engine', { line: line.slice(0, 200) });
      }
    }
  }

  private rejectAll(err: Error): void {
    for (const [id, { reject, timer }] of this.pending) {
      clearTimeout(timer);
      reject(err);
    }
    this.pending.clear();
  }

  private findEngineEntry(): string | null {
    // pfaa_bridge.py is the ONLY valid entry point — it has the stdin/stdout loop.
    // framework.py defines classes but has no bridge protocol.
    const candidates = [
      // 1. Same directory as this package (pfaa-cli/pfaa_bridge.py)
      join(this.config.workingDir, 'pfaa_bridge.py'),
      // 2. Relative to enginePath (when cwd is project root)
      join(this.config.enginePath, 'pfaa-cli', 'pfaa_bridge.py'),
      // 3. One level up from cwd (pfaa-cli/../pfaa-cli/pfaa_bridge.py)
      join(this.config.enginePath, 'pfaa_bridge.py'),
      // 4. Resolve from this file's package location
      new URL('../../../pfaa_bridge.py', import.meta.url).pathname,
    ];
    for (const path of candidates) {
      if (existsSync(path)) return path;
    }
    return null;
  }
}

/**
 * Create a bridge with sensible defaults for the Aussie Agents engine.
 */
export function createBridge(overrides: Partial<BridgeConfig> = {}): PFAABridge {
  const defaults: BridgeConfig = {
    pythonPath: process.env['PFAA_PYTHON_PATH'] || 'python3.15',
    enginePath: join(process.cwd(), '..'),
    workingDir: process.cwd(),
    timeoutMs: 120_000,
    maxConcurrent: 8,
    startupTimeoutMs: 30_000,
  };

  return new PFAABridge({ ...defaults, ...overrides });
}
